File: app/abstract/print_matrix.py
# coding=utf-8
# author: Xiguang Liu<g10guang@foxmail.com>
# 2018-04-28 18:11
# 题目描述：https://www.nowcoder.com/practice/9b4c81a02cd34f76be2659fa0d54342a?tpId=13&tqId=11172&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking


class Solution:
    def __init__(self):
        self.threshold = None
        self.matrix = None
        self.row = 0
        self.col = -1
        self.result = []
        self.resultLen = 0

    # matrix类型为二维列表，需要返回列表
    def printMatrix(self, matrix):
        """
        思路：定义四个方向的走向 → ↓ ← ↑ 循环遍历，
        给每一个方向定义一个尽头然后更新方向的信息
        """
        self.threshold = {'right': len(matrix[0]) - 1, 'down': len(matrix) - 1, 'left': 0, 'up': 1}
        self.matrix = matrix
        self.resultLen = len(matrix) * len(matrix[0])
        while self.resultLen > len(self.result):
            self.right()
            self.down()
            self.left()
            self.up()
        return self.result

    def right(self):
        if self.resultLen == len(self.result):
            return
        while self.col < self.threshold['right']:
            self.col += 1
            self.result.append(self.matrix[self.row][self.col])
        self.threshold['right'] -= 1

    def down(self):
        if self.resultLen == len(self.result):
            return
        while self.row < self.threshold['down']:
            self.row += 1
            self.result.append(self.matrix[self.row][self.col])
        self.threshold['down'] -= 1

    def left(self):
        if self.resultLen == len(self.result):
            return
        while self.col > self.threshold['left']:
            self.col -= 1
            self.result.append(self.matrix[self.row][self.col])
        self.threshold['left'] += 1

    def up(self):
        if self.resultLen == len(self.result):
            return
        while self.row > self.threshold['up']:
            self.row -= 1
            self.result.append(self.matrix[self.row][self.col])
        self.threshold['up'] += 1


import unittest


class TestCase(unittest.TestCase):
    def setUp(self):
        self.s = Solution()

    def test_1(self):
        matrix = [[1]]
        expect = [1]
        r = self.s.printMatrix(matrix)
        self.assertEqual(r, expect)

    def test_2(self):
        matrix = [[1, 2], [3, 4]]
        expect = [1, 2, 4, 3]
        r = self.s.printMatrix(matrix)
        self.assertEqual(r, expect)

    def test_3(self):
        matrix = [[1], [2], [3], [4], [5]]
        expect = [1, 2, 3, 4, 5]
        r = self.s.printMatrix(matrix)
        self.assertEqual(r, expect)

File: app/abstract/print_matrix.go
package main

/**
 * 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
 *
 * @param matrix int整型二维数组
 * @return int整型一维数组
 */
// https://www.nowcoder.com/practice/9b4c81a02cd34f76be2659fa0d54342a?tpId=13&tqId=11172&tPage=1&rp=1&ru=%2Fta%2Fcoding-interviews&qru=%2Fta%2Fcoding-interviews%2Fquestion-ranking
func printMatrix(matrix [][]int) []int {
	lenr := len(matrix)
	if lenr == 0 {
		return nil
	}
	lenc := len(matrix[0])
	if lenc == 0 {
		return nil
	}
	result := make([]int, 0, lenr*lenc)
	minr := 0
	maxr := lenr - 1
	minc := 0
	maxc := lenc - 1
	for minr <= maxr && minc <= maxc {
		// 向右遍历
		for y := minc; y <= maxc; y++ {
			result = append(result, matrix[minr][y])
		}
		// 向下遍历
		for x := minr + 1; x <= maxr; x++ {
			result = append(result, matrix[x][maxc])
		}
		// 向左遍历
		for y := maxc - 1; y >= minc && minr != maxr; y-- {
			result = append(result, matrix[maxr][y])
		}
		// 向上遍历
		for x := maxr - 1; x > minr && minc != maxc; x-- {
			result = append(result, matrix[x][minc])
		}
		minr++
		maxr--
		minc++
		maxc--
	}
	return result
}
